sw/host/opentitansession/src/main.rs

```diff
@@ -96,6 +113,119 @@ fn parse_command_line(opts: Opts, mut args: ArgsOs) -> Result<Opts> {
+    let mut args = args_os();
+
+    let mut child = Command::new(args.nth(0).unwrap()) // Same executable
+        .arg("--child") // Add argument to let the new process know it is the daemon child
+        .args(args) // Propagate all existing arguments: --interface, etc.
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    match serde_json::from_reader::<&mut ChildStdout, Result<SessionStartResult, String>>
+        (child.stdout.as_mut().unwrap())
+    {
+        Ok(Ok(result)) => {
+            // Create a pid file corresponding to the requested TCP port.
+            let path = run_file_fn(result.port);
+            File::create(path)?.write_all(format!("{}\n", child.id()).as_bytes())?;
+            Ok(Box::new(result))
+        }
+        Ok(Err(e)) => bail!(e),
+        Err(e) => bail!("Child process failed to start: {}", e),
+    }
+}
+
+// This method runs in the daemon child.  It will instantiate SessionHandler to bind to a
+// socket, then report the chosen port number to the parent process by means of a serialized
+// `SessionStartResult` sent through the stdout anonymous pipe, and finally enter an infnite
+// loop, processing connections on that socket
+fn session_child(listen_port: Option<u16>, backend_opts: &backend::BackendOpts) -> Result<()> {
+    let transport = backend::create(backend_opts, "null")?;
+    let mut session = SessionHandler::init(&transport, listen_port)?;
+    // Instantiation of Transport backen, and binding to a socket was successful, now go
+    // thought the process of making this process a daemon, disconeected from the
+    // terminal that was used to start it.
+
+    env::set_current_dir(&Path::new("/"))?;
+
+    // Close stdin and stderr, which would likely be connected to the tty (terminal)
+    // used to run opentitansession.
+    {
+        let devnull_file = File::open("/dev/null")?;
+        // devnull_fd is valid only as long as devnull_file remains in scope.
+        let devnull_fd: RawFd = devnull_file.as_raw_fd();
+        dup2(devnull_fd, 0)?;
+        dup2(devnull_fd, 2)?;
+    }
+    // After severing any stdio/err connections to the controlling terminal of the
+    // parent, we can now establish new Unix "session" for this process, which will not
+    // be connected to any terminal.  This means that this daemon will not be killed by
+    // SIGHUP, in case the terminal that was used for running `session start` is later
+    // closed.
+    setsid()?;
+
+    // Report startup success to parent process.
+    serde_json::to_writer::<io::Stdout, Result<SessionStartResult, String>>(
+        io::stdout(),
+        &Ok(SessionStartResult {
+            port: session.get_port(),
+        }),
+    )?;
+    io::stdout().flush()?;
+
+    // Closing the standard output pipe is the signal to the parent process that this
+    // child has started up successfully.  We close the pipe indirectly, by replacing
+    // file descriptor 1 with one pointing to /dev/null.
+    dup2(0, 1)?;
```

Comment:
Couldn't you just send e.g. a zero byte instead? Or any other kind of byte that won't show up as JSON.

Reply:
Even if I sent a zero byte, I do not want this child process to hold on to a file descriptor of an anonymous pipe, for which the parent process that held the other end has long since terminated.  Any accidental write on fd 1 would result in immediate termination with SIGPIPE.

Reply:
SIGPIPE is so stupid. Let's leave it as is, I guess.